### ISSUE TITLE
Proposal: Drop Hydra vocabulary in favour of JSON [Hyper-]Schema

### ISSUE THREAD

**Issue by lukeshope** [urbanobservatory/standards#13]
Now that I have actually started trying to use some of the things we've been discussing in a real world project, I have concerns that Hydra as a hypermedia doesn't provide enough of the descriptive power we need. 

90% of our API functionality is providing access to sensor data through some sensible ontologies, but there are several instances where clients will need to go beyond following links, and instead compose IRIs to describe the filters, constraints, pages, etc. they want.

We know as a minimum that our hypermedia controls need to allow:
  * Filtering collections based on properties (#10)
  * Partial views on collections based on pagination or time slices (#6)

I have no doubt more complex scenarios will arise. Especially if people implement API endpoints for sensor management.

To be clear, there is JSON Schema (generally describing the shape of a document), and JSON Hyper-Schema (describing how to interact with an API). I am suggesting implementations MUST use link descriptions (LDO) from Hyper-Schema, but MAY also use bits of JSON Schema (which would be useful for allowing client-side validation for example).

We would use:
* **JSON-LD** to reference our data to common vocabularies and ontologies, as a way of providing interoperability between the observatories and with other linked data APIs, RDF, etc.
* **JSON Hyper Schema** to describe how to search and filter our APIs.
* **JSON Schema** if people want to implement client-side validation, or to constrain what is allowed in their JSON-LD

**Advantages**
* A basic implementation would probably only need to use two schemas, one for pagination and one for filtering. The pagination one I've already done below :-)
* JSON [Hyper-]Schema is an active standard, much more so than Hydra seems to be. The latest version [`draft-08` is in final review](https://json-schema.org/work-in-progress).
* JSON Schema is near-identical to OpenAPI (formerly Swagger), which can be used in all sorts of testing and documentation frameworks
* Extensibility. Almost any API can be described using JSON [Hyper-]Schema. It can even be used to [validate JSON-LD documents](https://github.com/json-ld/json-ld.org/blob/master/schemas/jsonld-schema.json).
* Automated documentation using [Doca](https://blog.cloudflare.com/cloudflares-json-powered-documentation-generator/).
* Greater likelihood of libraries that work out of the box for us, such as [ajv](https://github.com/epoberezkin/ajv).

**Disadvantages**
* Complexity
* Learning curve

## Crash course in JSON Hyper-Schema

I've taken an example from the [9.5 Collections in the Hyper-Schema standard](https://json-schema.org/work-in-progress/WIP-jsonschema-hyperschema.html#rfc.section.9.5), and tried to make it more applicable to us. Assuming a request like `GET http://example.webof.uk/sensor/` to obtain a list of all the sensors, the response would look something like:
```json
Link: <http://example.webof.uk/schemas/v1.0.0/sensor-collection.json>; rel="describedBy"

{
  "elements": [
    {"@id": "http://example.webof.uk/sensor/sensor-a"},
    {"@id": "http://example.webof.uk/sensor/sensor-b"}
  ],
  "meta": {
    "current": { "offset": 0, "limit": 2 },
    "next": { "offset": 3, "limit": 2 }
  }
}
```
The elements in the body could of course contain more properties, or could be left as `@id` with the onus on the clients to dereference. 

The schema to describe the interaction patterns associated with the above document would be obtained with `GET http://example.webof.uk/schemas/v1.0.0/sensor-collection.json`, with a semantic version number.
```json
{
  "properties": {
    "elements": {
      "type": "array",
      "items": { "$ref": "#/definitions/jsonld-id" }
    },
    "meta": {
      "type": "object",
      "properties": {
        "prev": {"$ref": "#/definitions/pagination"},
        "current": {"$ref": "#/definitions/pagination"},
        "next": {"$ref": "#/definitions/pagination"}
      }
    }
  },
  "links": [
    {
      "rel": "self",
      "href": "sensor{?offset,limit}",
      "templateRequired": ["offset", "limit"],
      "templatePointers": {
        "offset": "/meta/current/offset",
        "limit": "/meta/current/limit"
      },
      "targetSchema": {"$ref": "#"}
    }, {
      "rel": "prev",
      "href": "sensor{?offset,limit}",
      "templateRequired": ["offset", "limit"],
      "templatePointers": {
        "offset": "/meta/prev/offset",
        "limit": "/meta/prev/limit"
      },
      "targetSchema": {"$ref": "#"}
    }, {
      "rel": "next",
      "href": "sensor{?offset,limit}",
      "templateRequired": ["offset", "limit"],
      "templatePointers": {
        "offset": "/meta/next/offset",
        "limit": "/meta/next/limit"
      },
      "targetSchema": {"$ref": "#"}
    }
  ],
  "definitions": {
    "jsonld-id": {
      "type": "object",
      "properties": {
        "@id": { "type": "string", "format": "uri" }
      },
      "required": [ "@id" ]
    },
    "pagination": {
      "type": "object",
      "properties": {
        "offset": { "type": "integer", "minimum": 0, "default": 0 },
        "limit": { "type": "integer", "minimum": 1, "maximum": 100, "default": 10 }
      }
    }
  }
}
```
This describes all of the links associated with the collection, those being the previous/next/current pages. It also describes the maximum number of items that can be requested per page etc., and the default values. These IRIs for the pages are generate using the IRI templates with pointers to the data within the document, in the `meta` object.

You can try the two above snippets in the [JSON Schema validator](https://www.jsonschemavalidator.net/).

The client would, with the above two documents, be able to infer the links available, including recognising that `prev` was absent from the `meta` object in the document, so no previous page link is available. This would provide the following:
```json
[
  {
    "contextUri": "https://example.webof.uk/sensor",
    "contextPointer": "",
    "rel": "self",
    "targetUri": "https://example.webof.uk/sensor?offset=0&limit=2",
    "attachmentPointer": ""
  },
  {
    "contextUri": "https://example.webof.uk/sensor",
    "contextPointer": "",
    "rel": "next",
    "targetUri": "https://example.webof.uk/sensor?offset=3&limit=2",
    "attachmentPointer": ""
  }
]
```

**Comment by SiBell** [comment]
Well the advantages sound good. I'm all for automated documentation. You're not wrong about the learning curve though. There's a series of blogs on https://apisyouwonthate.com that give a good introduction to JSON hyper-schema ([Part 1](https://apisyouwonthate.com/blog/getting-started-with-json-hyper-schema), [Part 2](https://apisyouwonthate.com/blog/getting-started-with-json-hyper-schema-part-2), [Part 3](https://apisyouwonthate.com/blog/common-hypermedia-patterns-with-json-hyper-schema)).

Could do with working through another real-world example, or two, to get a sense of just how easy this is to implement in practise.

It worries me slightly that doca hasn't been updated in a while, nor does it have many weekly downloads from npm in comparison to something like [ReDoc](https://www.npmjs.com/package/redoc).

Also is the line: `Link: <http://example.webof.uk/schemas/v1.0.0/sensor-collection.json>; rel="describedBy"` actually part of the response body?

**Comment by lukeshope** [comment]
Absolutely. I will try to hammer out a couple more examples over the coming days, perhaps focusing on our other big hypermedia need: filtering and searching.

I take your point about doca. Instead you could use ReDoc, which uses OpenAPI, though there are a few subtle differences between OpenAPI and JSON Schema. Phil has [created a converter](https://phil.tech/api/2018/03/30/openapi-and-json-schema-divergence/). I'm not sure Redoc would pick up the JSON Hyper-Schema elements that describe interactions, probably only the JSON Schema elements that describe validation and structure. Either way, we should be able to put something together that works.

The `Link` part is a header field in the response sorry, not part of the body.

**Comment by EttoreHector** [comment]
Guys, thank you so much for this. You are surely a couple of levels above me when it comes to linked-data technologies and standards. A couple of examples that relate with typical use-cases in the observatories would be really great!

**Comment by lukeshope** [comment]
Cheers, I am working on some examples and some tooling. Draft PR is #16 where I've tried to explain what I'm doing and why. More to follow next week!

:-)

**Comment by SiBell** [comment]
I've been looking into this a bit more. My thoughts:

- JSON Schema makes a lot of sense. Especially now it's better aligned with OpenAPI. Here's a good [article](https://apisyouwonthate.com/blog/openapi-v31-and-json-schema-2019-09) on the convergence.
- There's no shortage of [OpenAPI tools](https://openapi.tools/). In particular [stoplight studio](https://stoplight.io/studio/) looks like a fantastic user-friendly way of generating JSON Schemas, an OpenAPI document, and building and serving the documentation to users. The only downsides I can see are that it doesn't appear to automatically integrate with your actual server code (e.g. a Node.js express.js app). E.g. if you change a JSON schema file in your express.js app, then you'd need to manually copy and paste the updated JSON schema into stoplight. It's not clear if stoplight fully supports *JSON Schema 2019-09* and *OpenAPI v3.1*  yet, but you'd imagine it will be one of the first to do so.
- I'm still baffled by JSON Hyper-Schema, other than the [official docs](https://json-schema.org/draft/2019-09/json-schema-hypermedia.html) and the [apisyouwonthate.com](https://apisyouwonthate.com) blogs, there's so little extra info or tooling for it. It's still not super clear to me how a front-end app would be able to use it to paginate through a resource.
